logging/loggeroption.go

```diff
@@ -0,0 +1,193 @@
+
+// ContextFunc is a function that will be called to obtain a context.Context for the
+// WriteLogEntries RPC executed in the background for calls to Logger.Log. The
+// default is a function that always returns context.Background. The second return
+// value of the function is a function to call after the RPC completes.
+//
+// The function is not used for calls to Logger.LogSync, since the caller can pass
+// in the context directly.
+//
+// This option is EXPERIMENTAL. It may be changed or removed.
+func ContextFunc(f func() (ctx context.Context, afterCall func())) LoggerOption {
+	return contextFunc(f)
+}
+
+type contextFunc func() (ctx context.Context, afterCall func())
+
+func (c contextFunc) set(l *Logger) { l.ctxFunc = c }
+
+// SourceLocationPopulation is the flag controlling population of the source location info
+// in the ingested entries. This options allows to configure automatic population of the
+// SourceLocation field for all ingested entries, entries with DEBUG severity or disable it.
+// Note that enabling this option can decrease execution time of Logger.Log and Logger.LogSync
+// by the factor of 2 or larger.
+// The default disables source location population.
+//
+// This option is not used when an entry is created using ToLogEntry.
+func SourceLocationPopulation(f int) LoggerOption {
+	return sourceLocationOption(f)
+}
+
+const (
+	// DoNotPopulateSourceLocation is default for clients when WithSourceLocation is not provided
+	DoNotPopulateSourceLocation = 0
+	// PopulateSourceLocationForDebugEntries is set when WithSourceLocation(PopulateDebugEntries) is provided
+	PopulateSourceLocationForDebugEntries = 1
+	// AlwaysPopulateSourceLocation is set when WithSourceLocation(PopulateAllEntries) is provided
+	AlwaysPopulateSourceLocation = 2
+)
+
+type sourceLocationOption int
+
+func (o sourceLocationOption) set(l *Logger) {
+	if o == DoNotPopulateSourceLocation || o == PopulateSourceLocationForDebugEntries || o == AlwaysPopulateSourceLocation {
+		l.populateSourceLocation = int(o)
+	}
+}
+
+// PartialSuccess sets the partialSuccess flag to true when ingesting a bundle of log entries.
+// See https://cloud.google.com/logging/docs/reference/v2/rest/v2/entries/write#body.request_body.FIELDS.partial_success
+// If not provided the partialSuccess flag is set to false.
+func PartialSuccess() LoggerOption {
+	return &partialSuccessOption{}
+}
+
+type partialSuccessOption struct{}
+
+func (o *partialSuccessOption) set(l *Logger) {
+	l.partialSuccess = true
+}
+
+// RedirectToStdout configures the logger to print log entries to stdout as Jsonified strings instead of ingesting them
+// to Cloud Logging. See https://cloud.google.com/logging/docs/structured-logging#special-payload-fields for Json format
+// of the printed strings. Use this option when need to delegate log entries ingestion to out-of-process logging agent.
+func RedirectToStdout() LoggerOption {
```

Comment:
I think these two options should be combined to one. Something like `LogJSON(io.Writer)`. If the writer writter is nil default it to Stderr or something. That way there is more flexibility for users that might want to send the logs to a file for instance.

Reply:
There were a few considerations to make it this way. I thought that making `RedirectToStdout` is more verbose than `Redirect(os.Stdout)`. Allowing any _writer_ opens a door for misuse and confusion because the format that the library uses to print the logs is supported **only** by GCP logging agents. Other uses will lead to logs not to be ingested to Cloud Logging which in turn defeats the purpose of the library.
The name such as "LogJSON" is misleading due the same considerations. While the format is JSON, it uses special fields that guarantee a specific behavior and only when logs are printed to standard output.